### PR TITLE
Divide comm transport layer into client and server

### PR DESCRIPTION
**Pull request by sebtomba**
## Overview
Divide the transport layer into a client and a server. Refactoring of the `TcpTransportLayer` into several traits and classes for a cleaner code structure.

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-2929

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### CODE REVIEW

**Comment by sebtomba** [comment]
bors try

**Comment by sebtomba** [comment]
bors try

**Comment by sebtomba** [comment]
bors try

**Comment by rabbitonweb** [comment]
reviewed with @sebtomba yesterday online, waiting for discussed changes

**Comment by sebtomba** [comment]
@sebtomba Changed. However, I joined all `GrpcTransportProtocol`
traits into `GrpcTransportClient`

File: comm/src/main/scala/coop/rchain/comm/transport/GrpcTransportClient.scala [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        _ <- s.connections.get(peer) match {
+              case Some(c) =>
+                log
+                  .debug(s"Disconnecting from peer ${peer.toAddress}")
+                  .map(kp(Try(c.shutdown())))
+                  .void
+              case _ => Task.unit // ignore if connection does not exists already
+            }
+      } yield s.copy(connections = s.connections - peer)
+    }
```
Comment (rabbitonweb) [inline comm/src/main/scala/coop/rchain/comm/transport/GrpcTransportClient.scala]: consider
```
    cache.modify { s =>  s.connections.get(peer) match {
              case Some(c) =>
                log
                  .debug(s"Disconnecting from peer ${peer.toAddress}")
                  .map(kp(Try(c.shutdown())))
                  .void
              case _ => Task.unit // ignore if connection does not exists already
            }
      } map s.copy(connections = s.connections - peer)
```
instead

File: comm/src/main/scala/coop/rchain/comm/transport/GrpcTransportClient.scala [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        peers <- cache.read.map(_.connections.keys.toSeq)
+        _     <- log.info("Sending shutdown message to all peers")
+        _     <- innerBroadcast(peers, msg)
+        _     <- log.info("Disconnecting from all peers")
+        _     <- Task.gatherUnordered(peers.map(disconnect))
+      } yield ()
+
+    cache.read.flatMap { s =>
+      if (s.shutdown) Task.unit
+      else shutdownQueue *> sendShutdownMessages
```
Comment (rabbitonweb) [inline comm/src/main/scala/coop/rchain/comm/transport/GrpcTransportClient.scala]: use `>>` instead 
1. stack saftey
2. order of execution!

Comment (sebtomba) [reply comm/src/main/scala/coop/rchain/comm/transport/GrpcTransportClient.scala]: You made my day! :)

**Comment by sebtomba** [comment]
@rabbitonweb changed as requested. Please review.

**Pull request merged**